src/cunumeric/sort/sort.cu

```diff
@@ -0,0 +1,691 @@
+                                         samples.ptr(0),
+                                         samples.ptr(0) + num_global_samples,
+                                         init_sample,
+                                         SampleComparator<VAL>());
+  size_t num_usable_samples = lower_bound - samples.ptr(0);
+
+  // select splitters / positions based on samples (on device)
+  const size_t num_splitters = num_ranks - 1;
+  auto split_positions       = create_buffer<size_t>(num_splitters, Memory::Z_COPY_MEM);
+  {
+    const size_t num_blocks = (num_splitters + THREADS_PER_BLOCK - 1) / THREADS_PER_BLOCK;
+    VAL init_value          = std::numeric_limits<VAL>::max();
+    extract_split_positions<<<num_blocks, THREADS_PER_BLOCK, 0, stream>>>(
+      local_sorted.values.ptr(0),
+      volume,
+      samples.ptr(0),
+      num_usable_samples,
+      split_positions.ptr(0),
+      num_splitters,
+      my_rank);
+  }
+
+  // need to sync as we share values in between host/device
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+
+  // collect sizes2send, send to rank i: local_sort_data from positions  split_positions[i-1],
+  // split_positions[i] - 1
+  auto size_send = create_buffer<size_t>(num_ranks, Memory::Z_COPY_MEM);
+  {
+    size_t last_position = 0;
+    for (size_t rank = 0; rank < num_ranks - 1; ++rank) {
+      size_t cur_position = split_positions[rank];
+      size_send[rank]     = cur_position - last_position;
+      last_position       = cur_position;
+    }
+    size_send[num_ranks - 1] = volume - last_position;
+  }
+
+  // need to sync as we share values in between host/device
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+
+  // all2all exchange send/receive sizes
+  auto size_recv = create_buffer<size_t>(num_ranks, Memory::Z_COPY_MEM);
+  CHECK_NCCL(ncclGroupStart());
+  for (int r = 0; r < num_ranks; r++) {
+    CHECK_NCCL(ncclSend(size_send.ptr(r), 1, ncclUint64, r, *comm, stream));
+    CHECK_NCCL(ncclRecv(size_recv.ptr(r), 1, ncclUint64, r, *comm, stream));
+  }
+  CHECK_NCCL(ncclGroupEnd());
+
+  // need to sync as we share values in between host/device
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+
+  // allocate merge targets, data transfer...
+  std::vector<SortPiece<VAL>> merge_buffers(num_ranks);
+
+  for (int i = 0; i < merge_buffers.size(); ++i) {
+    // align buffer to allow data transfer of 16byte blocks
+    auto recv_size_aligned   = get_aligned_size(size_recv[i] * sizeof(VAL));
+    auto buf_size            = (recv_size_aligned + sizeof(VAL) - 1) / sizeof(VAL);
+    merge_buffers[i].values  = create_buffer<VAL>(buf_size, Memory::GPU_FB_MEM);
+    merge_buffers[i].indices = create_buffer<int64_t>(argsort ? buf_size : 0, Memory::GPU_FB_MEM);
+    merge_buffers[i].size    = size_recv[i];
+  }
+  size_t send_pos = 0;
+  CHECK_NCCL(ncclGroupStart());
+  for (int r = 0; r < num_ranks; r++) {
+    CHECK_NCCL(ncclSend(local_sorted.values.ptr(send_pos),
```

Comment:
there's a buffer overrun in this line: the chunk you send to rank `num_ranks - 1` may go over the size of `local_sorted` because of the alignment.

Reply:
Is there a way to solve this without making another copy? I could not find any information on how alignment is affecting NCCL, I just had this from the 'unique' code. Should I just ignore the alignment here and transfer exactly?

Reply:
I don't know whether the bug was fixed in recent NCCL versions, but there were two major issues in NCCL <= 2.8: First, transfers with unaligned allocations are at least an order of magnitude slower than those with aligned ones; Second, it wasn't allowed to have some of the participants not send any data, i.e., every pair of ranks must exchange non-empty data. My all-to-all code was written to avoid these issues.

I suggest you just make copies in buffers of aligned sizes for the first PR and later optimize the memory usage. The alignment is required for both the size of the transfer and the starting address of the allocation, so copies don't seem avoidable.

Reply:
I added additional send-buffers if needed